algorithms/src/snark/varuna/ahp/indexer/indexer.rs

```diff
@@ -224,27 +227,35 @@ impl<F: PrimeField, SM: SNARKMode> AHPForR1CS<F, SM> {
         result
     }
 
+    /// Evaluate the index polynomials for this constraint system at the given point.
+    /// Return the LinearCombination of the index polynomials and the sum of the evaluations.
     pub(crate) fn evaluate_index_polynomials(
         state: IndexerState<F>,
         id: &CircuitId,
         point: F,
-    ) -> Result<impl Iterator<Item = F>, AHPError> {
-        let labels = [["a"], ["b"], ["c"]];
-        let mut evals = [
-            (state.a_arith, state.non_zero_a_domain),
-            (state.b_arith, state.non_zero_b_domain),
-            (state.c_arith, state.non_zero_c_domain),
-        ]
-        .into_iter()
-        .zip_eq(&labels)
-        .flat_map(|((evals, domain), label)| {
-            let labels = Self::index_polynomial_labels(label, std::iter::once(id));
+        mut combiners: impl Iterator<Item = F>,
+    ) -> Result<(LinearCombination<F>, F)> {
+        let mut lc = LinearCombination::empty("circuit_check");
+        let mut all_evals = Vec::with_capacity(3);
+        let mut sum = F::zero();
+        for (evals, domain, label) in [
+            (state.a_arith, state.non_zero_a_domain, "a"),
+            (state.b_arith, state.non_zero_b_domain, "b"),
+            (state.c_arith, state.non_zero_c_domain, "c"),
+        ] {
+            let labels = Self::index_polynomial_labels_single(label, id);
             let lagrange_coefficients_at_point = domain.evaluate_all_lagrange_coefficients(point);
-            labels.zip_eq(evals.evaluate(&lagrange_coefficients_at_point).unwrap())
-        })
-        .collect::<Vec<_>>();
-        evals.sort_by(|(l1, _), (l2, _)| l1.cmp(l2));
-        Ok(evals.into_iter().map(|(_, eval)| eval))
+            let evals_at_point = evals.evaluate(&lagrange_coefficients_at_point)?;
+            ensure!(labels.len() == evals_at_point.len());
+            all_evals.push(labels.into_iter().zip_eq(evals_at_point.into_iter()));
+        }
+        let sorted_evals = all_evals.into_iter().flatten().sorted_unstable_by(|(l1, _), (l2, _)| l1.cmp(l2));
+        for (label, eval) in sorted_evals {
+            let combiner = combiners.next().ok_or(anyhow!("No combiner left"))?;
```

Comment:
Is it safe for `combiners` to be longer than `sorted_evals`?

If not, we should add an `ensure!` on the lengths.